### DISCUSSION TITLE
Bug: Interface 'DayPickerSingleProps' incorrectly extends interface 'DayPickerBase'

### DISCUSSION THREAD

**Discussion by claviustales** [gpbl/react-day-picker#1455] [Support]
### Bug description

Compiling a project that uses TypeScript 4.6.4, get the following:

```
node_modules/react-day-picker/dist/types/DayPickerSingle.d.ts:6:18 - error TS2430: Interface 'DayPickerSingleProps' incorrectly extends interface 'DayPickerBase'.
  Types of property 'selected' are incompatible.
    Type 'Date | undefined' is not assignable to type 'Matcher | Matcher[]'.
      Type 'undefined' is not assignable to type 'Matcher | Matcher[]'.
```

### To reproduce

I could not reproduce on CodeSandbox.

**Comment by gpbl** [comment]
@claviustales please provide some context:

- some code to investigate :)
- are you migrating from v7?

**Reply by claviustales** [reply]
Hi, @gpbl.

As you intuited, I'm migrating from v7. I've followed the [guide](react-day-picker.js.org/guides/upgrading). I'm using _yarn_. The problem occurs when I run _yarn compile_.

About sharing coding: as shown on problem description, the issue is not on my code; it's on lib code.

It is simple to see the problem on lib code:

On [this file](https://github.com/gpbl/react-day-picker/blob/77bbb66c68a58b6915a23109efa8af09f0671c85/packages/react-day-picker/src/types/DayPickerBase.ts):
```
export interface DayPickerBase {
  ...
  selected?: Matcher | Matcher[];
  ...
}
```

On [this file](https://github.com/gpbl/react-day-picker/blob/77bbb66c68a58b6915a23109efa8af09f0671c85/packages/react-day-picker/src/types/DayPickerSingle.ts):
```
export interface DayPickerSingleProps extends DayPickerBase {
  ...
  selected?: Date | undefined;
  ...
}
```

The second interface inherits from the first interface and "redefine" the type of _selected_ property, that is not allowed by TypeScript.

**Reply by gpbl** [reply]
> The second interface inherits from the first interface and "redefine" the type of selected property, that is not allowed by TypeScript.

That should not be the case, as `Date` is a subset of `Matcher`.

- what does `yarn compile` do?
- did you remove the previous types?
- which version of typescript are you using?

**Reply by claviustales** [reply]
> That should not be the case, as `Date` is a subset of `Matcher`.

Because of that your statement, I've got the real problem:

In my project, [exactOptionalPropertyTypes](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes) compile option is turned on. In this case, `Date | undefined` is not a subset of `Matcher`, as `Matcher` does not include `undefined`.

**Reply by gpbl** [reply]
Oh that is interesting. I guess if we add `undefined` to https://github.com/gpbl/react-day-picker/blob/77bbb66c68a58b6915a23109efa8af09f0671c85/packages/react-day-picker/src/types/DayPickerBase.ts#L186

the problem should go away.

**Reply by claviustales** [reply]
Exactly. This would be my suggestion.
